cheatsheets/Protect_FileUpload_Against_Malicious_File.md

```diff
@@ -1,597 +1,138 @@
+- **Allowed users only can upload files**
+- **Validate the file type, don't trust the [Content-Type header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type)  as it can be spoofed**
+- **Change the file name to something generated by the application**
+- **Set a file name length limit. Restrict the allowed characters if possible**
+  - **Setting a character limit could deny the attacker from sending null bytes, adding secondary extensions, and performing XSS attacks if the file name is reflected**
+- **Set a file size limit**
+- **Store the files on a different server. If that's not possible, store them in a private directory**
+  - **In the case of public access to the files, ensure that proper permissions are set, file names are safe, and encoded**
+- **Run the file through an antivirus or a sandbox if available to validate that it doesn't contain malicious or bogus data**
+- **Ensure the usage of secure and updated libraries for your needs**
 
-Usually, when an application expect his user to upload a document, the application expect to receive a document for which the intended use will be for reading/printing/archiving. The document should not alter is content at opening time and should be in a final rendered state.
+## Contents
 
-The most common file types used to transmit *malicious code* into file upload feature are the following:
+- [File Upload Threats](#file-upload-threats)
+    - [Malware Upload](#malware-upload)
+    - [Public File Retrieval](#public-file-retrieval)
+- [File Upload Protection](#file-upload-protection)
+    - [Extension Validation](#extension-validation)
+        - [Whitelist Extensions](#whitelist-extensions)
+        - [Blacklist Extensions](#blacklist-extensions)
+    - [User Permissions](#user-permissions)
+    - [Content-Type Validation](#content-type-validation)
+    - [File Name Sanitization](#file-name-sanitization)
+    - [File Storage Location](#file-storage-location)
+    - [System Permissions](#system-permissions)
+    - [Upload and Download Limits](#upload-and-download-limits)
+- [Java Code Snippets](#java-code-snippets)
 
-- Microsoft Office document: Word/Excel/PowerPoint using [VBA Macro](https://en.wikipedia.org/wiki/Visual_Basic_for_Applications) and [OLE package](https://en.wikipedia.org/wiki/Object_Linking_and_Embedding).
-- Adobe PDF document: Insert malicious code as attachment.
-- Images: Malicious code embedded into the file or use of binary file with image file extension.
+## File Upload Threats
 
-# Approaches
+In order to assess and know exactly what controls to implement, knowing what you're facing is essential to protect your assets. The following sections will hopefully showcase the risks accompanying the file upload functionality.
 
-Based on this context, the goals here are:
+### Malware Upload
```

Comment:
Also need to have a section about dealing with the the contents of files, which could be:

* Personal data (GDPR issues)
* Copyrighted (piracy)
* Illegal (CSE, stolen data, etc)
* Reputationaly damaging (especially if publicly available)
* Dangerous (macros, CSV formula injection, XXE payloads, etc)

Reply:
So I was considering that part, then again, how can they cover it? I discussed the last point, the dangerous files.

Reply:
It'll depend on exactly what the site is doing with uploaded files, but you could have stuff like:

* A process for users to report illegal/inappropriate files
* A process for copyright owners to report abuse
* Manual review of files (before they're made public)?
* Automated review (filename/keyword analysis, hash databases, image/video processing, etc)

Testing for Macros/CSV formula/XXE is going to be harder - it'll vary depending on the exact type of file